### PR TITLE
[Docs] Add an antipattern to the testing documentation page

### PR DESCRIPTION
**Pull request by peter-csala**
### The issue or feature being addressed
 - #1558 
 - #1549 

### Details on the issue fix or feature implementation
- I've tried to migrate the V7 example codes to V8 but they are wrong 

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [ ]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

**Comment by peter-csala** [comment]
Hey @martintmk Can you please help me by showing how to mock a pipeline?

File: docs/advanced/testing.md [hunk 1]
```diff
@@ -61,3 +61,76 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
 // ...
 ```
 <!-- endSnippet -->
+
+
+## Patterns and anti-patterns
+
+Over the years, many developers have used Polly in various ways. Some of these recurring patterns may not be ideal. This section highlights the recommended practices and those to avoid.
+
+### 1 - Recreating the policies inside unit tests
```
Comment (martintmk) [inline docs/advanced/testing.md:70]:
```suggestion
### 1 - Recreating the strategies inside unit tests
```

File: docs/advanced/testing.md [hunk 2]
```diff
@@ -61,3 +61,76 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+
+Imagine that you have a class that receives a `ResiliencePipeline<T>` as a consturtor parameter for testability.
```
Comment (martintmk) [inline docs/advanced/testing.md:72]:
```suggestion
Imagine that you have a class that receives a `ResiliencePipeline<T>` as a constructor's parameter for testability.
```

File: docs/advanced/testing.md [hunk 3]
```diff
@@ -61,3 +61,76 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+
+❌ DON'T
+
+Recreate the strategey inside the unit test:
```
Comment (martintmk) [inline docs/advanced/testing.md:76]:
```suggestion
Recreate the strategy inside the unit test:
```

File: docs/advanced/testing.md [hunk 4]
```diff
@@ -61,3 +61,76 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+
+- If you want to test your to-be-decorated code like there is no strategy applied
+
+<!-- snippet: testing-pattern-1-1 -->
+```cs
+var timeoutMock = ResiliencePipeline<HttpResponseMessage>.Empty;
+```
+<!-- endSnippet -->
+
+- If you want to verify that your code handles success cases as expected
```
Comment (martintmk) [inline docs/advanced/testing.md:113]:
```suggestion
- If you want to verify that your code handles success cases as expected:
```

**Comment by martintmk** [comment]
> Hey @martintmk Can you please help me by showing how to mock a pipeline?

Currently, there is no straightforward way to do so. You need to create a custom fake strategy, add it to `ResiliencePipelineBuilder`, build the pipeline and use it.

File: docs/advanced/testing.md (hunk 4 above)
Comment (martintmk) [reply docs/advanced/testing.md:113]: Not directly, if there is a desire to do it we might introduce some API into Polly.Testing package.

Something similar to:
https://github.com/App-vNext/Polly/blob/main/test/Polly.TestUtils/TestResilienceStrategy.cs

But I would rather wait for a feedback.

File: docs/advanced/testing.md [hunk 5]
```diff
@@ -61,3 +61,115 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+// Act
+var sut = new SUT(downstreamMock, timeout);
+await sut.RetrieveData();
+```
+<!-- endSnippet -->
+
+**Reasoning**:
+
+- Depending on the your strategy setup, the execution time might be greatly increased (from milliseconds to seconds).
+  - For example if you have a retry with 3 `MaxRetryAttempts` and 1 seconds `Delay`.
```
Comment (martincostello) [inline docs/advanced/testing.md:98]:
```suggestion
  - For example if you have a retry with 3 `MaxRetryAttempts` and 1 second's `Delay`.
```

File: docs/advanced/testing.md [hunk 6]
```diff
@@ -61,3 +61,115 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+- This violates one of the F.I.R.S.T testing principles, namely the **Fast** one.
```
Comment (martincostello) [inline docs/advanced/testing.md:99]: Let's link to what FIRST is for people who don't know it.

File: docs/advanced/testing.md [hunk 7]
```diff
@@ -61,3 +61,115 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+
+✅ DO
+
+Use `Empty` or mock `ExecuteAsync`:
```
Comment (martincostello) [inline docs/advanced/testing.md:103]:
```suggestion
Use `Empty` or mock `ExecuteAsync()`:
```

File: docs/advanced/testing.md [hunk 8]
```diff
@@ -61,3 +61,115 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+var sut = new SUT(downstreamMock, pipeline);
+await sut.RetrieveData();
+```
+<!-- endSnippet -->
+
+**Reasoning**:
+
+- Polly currently does not support direct mocking of `ResiliencePipeline`.
+  - As a workaround you can derive from the `ResilienceStrategy` class.
+  - You can expose an `ExecuteAsync` method with the same signature as you it inside the system-under-test.
```
Comment (martincostello) [inline docs/advanced/testing.md:170]: Not sure if my suggestion is what you intended, but what's there isn't right.

```suggestion
  - You can expose an `ExecuteAsync()` method with the same signature as how you use it inside the system-under-test.
```

File: docs/advanced/testing.md [hunk 9]
```diff
@@ -61,3 +61,115 @@ ResiliencePipelineDescriptor descriptor = pipeline.GetPipelineDescriptor();
...
+  - Then you add that strategy to a pipeline.
+- This approach allows you to define the expected behaviour for your resilience pipeline which meets your test case's needs.
+
+> [!NOTE]
+> The above presented solution is a workaround and later Polly might add support for this use case as well in the `Polly.Testing` package.
```
Comment (martincostello) [inline docs/advanced/testing.md:175]: I'm not sure we should have this here in our docs without a concrete issue somewhere asking for it. Unless you really really want to advocate for such a thing, I would just remove this note.

File: src/Snippets/Docs/Testing.cs [hunk 10]
```diff
@@ -58,4 +62,117 @@ public static void GetPipelineDescriptorGeneric()
...
+    {
+        /// <summary>
+        /// The method.
+        /// </summary>
+        /// <param name="token">The parameter.</param>
+        /// <returns>The result.</returns>
+        ValueTask<HttpResponseMessage> CallDownstreamAsync(CancellationToken token);
+    }
+
+    private class SUT
```
Comment (martincostello) [inline src/Snippets/Docs/Testing.cs:79]: Not everyone is familiar with _SUT_ for _System Under Test_ - can we rename this to something else?

File: src/Snippets/Docs/Testing.cs [hunk 11]
```diff
@@ -58,4 +62,117 @@ public static void GetPipelineDescriptorGeneric()
...
+        #endregion
+    }
+
+    public static async Task Pattern_11()
+    {
+        #region testing-pattern-1-1
+        // Arrange
+        var timeoutMock = ResiliencePipeline<HttpResponseMessage>.Empty;
+        var downstreamMock = Substitute.For<IDownstream>();
+        // setup the to-be-decorated method in a way that suits for your test case
```
Comment (martincostello) [inline src/Snippets/Docs/Testing.cs:118]:
```suggestion
        // Set up the to-be-decorated method in a way that is appropriate for your test case
```

**Comment by peter-csala** [comment]
I'm closing this issue because the suggested solution feels really unnatural and artifical.

**Pull request closed without merging**